server/chat-plugins/lottery.js

```diff
@@ -0,0 +1,248 @@
+'use strict';
+
+const LOTTERY_FILE = 'config/chat-plugins/lottery.json';
+
+/** @type {typeof import('../../lib/fs').FS} */
+const FS = require(/** @type {any} */('../../.lib-dist/fs')).FS;
+
+/** @type {{[roomid: string]: {maxWinners: number, name: string, markup: string, participants: {[ip: string]: string}}}} */
+let lotteries = {};
+try {
+	lotteries = require(`../../${LOTTERY_FILE}`);
+} catch (e) {
+	if (e.code !== 'MODULE_NOT_FOUND' && e.code !== 'ENOENT') throw e;
+}
+if (!lotteries || typeof lotteries !== 'object') lotteries = {};
+
+class Lottery {
+	/**
+     * @param {string} roomid
+     */
+	constructor(roomid) {
+		this.roomid = roomid;
+	}
+	/**
+	 * @param {number} maxWinners
+	 * @param {string} name
+	 * @param {string} markup
+	 */
+	create(maxWinners, name, markup) {
+		if (!lotteries[this.roomid]) {
+			lotteries[this.roomid] = {maxWinners, name, markup, participants: {}};
+			this.write();
+		}
+	}
+	write() {
+		FS(LOTTERY_FILE).writeUpdate(() => JSON.stringify(lotteries));
+	}
+	destroy() {
+		delete lotteries[this.roomid];
+		this.write();
+	}
+	/**
+     * @param {User} user
+     */
+	addUser(user) {
+		if (!lotteries[this.roomid].participants[user.latestIp]) {
```

Comment:
doesn't this mean that if I have two IPs, I'll be able to join twice under the same name?

Reply:
Fixed, but you can still just log in to another account, but nothing we could do about that, I think.